[Inductor max autotune] Multithreaded Precompilation

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115877
* #115814
* #115813
* #115655
* #115654
* #112861
* #114687
* #115270
* #115174
* #115072
* #114606
* #114319
* #114075
* #113932
* #113891
* #113959
* #113890
* #113570
* #113569
* __->__ #113558
* #113399
* #113366
* #113365

When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

kadeng committed:
[Inductor max autotune] Multithreaded Precompilation

When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aakhundov commented on torch/_inductor/select_algorithm.py:
```diff
@@ -774,7 +780,26 @@ def __call__(
+                executor.map(
+                    lambda c: c.precompile(),
+                    [c for c in choices if hasattr(c, "precompile")],
+                    timeout=60 * 15,
```
Maybe make this a constant?

kadeng replied:
Made it a parameter that's passed at construction..

aakhundov commented on torch/_inductor/select_algorithm.py:
```diff
@@ -774,7 +780,26 @@ def __call__(
+            try:
+                precompile(choices)
+            except TimeoutError:
+                log.warning("Precompilation phase took longer than timeout allowed. Continuing")
```
What happens in precompilation times out? Is the subsequent `make_benchmark_fn()(choices)` expected to run fine?

kadeng replied:
Yes, nothing bad happens in that case.

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation of CUDA source files can totally dominate the runtime required for the benchmarking done as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a possible on-disk sccache ).

Also it ensures that no unneccessary compilation and benchmarking steps are performed, which was peviously the case.

In sum, this can speed up the compilation of large models which have a lot of GEMMs considerably ( often by a factor of 20 or so ) if there are no limits to the number of Cutlass operations being considered for Autotuning.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented:
Moved to a (draft) feature branch, see https://github.com/pytorch/pytorch/pull/115919
